architecture/2024-11-authz-feature-spec.md

```diff
@@ -0,0 +1,390 @@
+
+Large enterprises need a flexible authorization model which enables them to tightly control a granular set of actions permitted on granular set of resources. To describe this scenario in more depth, a series of user stories will be described based on the diagram below.
+
+![](2024-11-authz-feature-spec/authz.png)
+
+#### User story 3 – Create resource group for environments
+
+As a platform engineer at a large enterprise, I need to create a resource group for a cloud environment and grant access to manage environments to my cloud infrastructure team. I have deleted the `default-environment` resource group.
+
+**User Experience**
+
+```bash
+# Create resource group for non-production environments
+rad group create non-prod-env
+# Assign Environment Administrator role to the cloud engineering user group
+rad role assignment create \
+  --assignee cloud-engineering@my-company.net \
+  --role env-admin \
+  --scope /planes/radius/MyCompany/resourceGroup/non-prod-env
+```
+
+**Result**
+
+1. An empty resource group is created called `non-prod-env`
+2. The `env-admin` role is granted to the cloud engineering group to manage environments within the `non-prod-env` resource group
+
+**Exceptions**
+
+The operation fails and informs the user interactively if:
+
+* The current user is not a member of the `radius-admin` or `group-admin` roles or the `cluster-admin` RoleBinding in the Kubernetes cluster
+* The resource group already exists
+* The role assignment already exists
+
+#### User story 4 – Create resource group for an application
+
+As a platform engineer at a large enterprise, I need to create a resource group for a new application and grant access to a development team. I only want the development team to use my company's resource types. I have deleted the `default-application` resource group.
+
+**User Experience**
+
+The user needs to (1) modify the Developer role so that only MyCompany.App resource types can be used and (2) grant the updated Developer role to the IdP group.
+
+```bash
+# Update the Developer role
+rad role definition update -f developer-role-definition.yaml
+```
+
+The contents of `developer-role-definition.yaml` is similar to:
+
+```yaml
+---
+name: developer
+description: Role to manage applications and application resources within a resource group
+actions:
+  # Grant permissions to CRUDL applications in the scoped resource group
+  - Applications.Core/applications/*
+  # Restrict using any resource type except those in the MyCompany.App namespace
+  - MyCompany.App/*
```

Comment:
Just thinking aloud, but it would be really fun to demo this with a `rad role-definition edit` (like `kubectl edit`).

I think that creates a really strong narrative about customization.